### PR TITLE
GH-113838: Add "Comparison to os.path" section to pathlib docs

### PR DESCRIPTION
**Pull request by barneygale**
* Issue: gh-113838

----
📚 Documentation preview 📚: https://cpython-previews--115926.org.readthedocs.build/

### CODE REVIEW

**Comment by barneygale** [comment]
Qs:

The pathlib docs are pretty beginner-friendly, so I went back to basics with the OOP vs procedural stuff. Is it too much?

Is the trailing slash bit clear enough?

Should the conclusion "pathlib is not a drop-in replacement" be softened?

File: Doc/library/pathlib.rst [hunk 1]
```diff
@@ -1682,27 +1682,57 @@ The patterns accepted and results generated by :meth:`Path.glob` and
...
+written in pure Python and is often slower, but rarely slow enough to matter.
+
+pathlib's path normalization is slightly more opinionated and consistent than
+:mod:`os.path`. For example, whereas :func:`os.path.abspath` eliminates
+"``..``" segments from a path, which may change its meaning if symlinks are
+involved, :meth:`Path.absolute` preserves these segments for greater safety.
+
+pathlib's path normalization may render it unsuitable for some applications:
+
+1. pathlib normalizes ``Path("foo/")`` to ``Path("foo")``, which changes a
```
Comment (SubaruArai) [inline Doc/library/pathlib.rst]: I forgot where I read it, but python docs (and I) dislike using `foo`/`bar`s, due to not being beginner-friendly.
Something like `Path("my_folder/")` might be better here.

File: Doc/library/pathlib.rst [hunk 2]
```diff
@@ -1682,27 +1682,57 @@ The patterns accepted and results generated by :meth:`Path.glob` and
...
+   path's meaning when supplied to various operating system APIs and
+   command-line utilities. Specifically, the absence of a trailing separator
+   allows the path to be resolved as either a file or directory, rather than a
+   directory only.
+2. pathlib normalizes ``Path("./foo")`` to ``Path("foo")``, which changes a
```
Comment (SubaruArai) [inline Doc/library/pathlib.rst]: Same as above, I prefer to avoid `foo`. In this case, maybe `file_or_folder`?

**Comment by SubaruArai** [comment]
Sorry, first time doing a review so what I've posted is quite a mess.

**Comment by barneygale** [comment]
Not at all, a great review IMO :]

**Comment by SubaruArai** [comment]
I'm thinking about changing line 42 (see os.path), maybe something like this should work?
I ain't sure what's the markup for "important" or "warning", but you get the idea.

After line 44, add:
```rst
.. important::
    pathlib is NOT a drop-in replacement for :mod:`os.path`. For more information, see :ref:`the comparison <comparison-to-the-os-and-os-path-modules>`.
```

**Comment by barneygale** [comment]
I think that might be giving it undue prominence. Suppressing leading `./` only really matters for executable search paths, and a great many users are either unaware of trailing slashes, or expect them to be removed for whatever reason.

**Comment by SubaruArai** [comment]
Maybe just add a link to the comparison?
```rst
.. seealso::
   For low-level path manipulation on strings, you can also use the
   :mod:`os.path` module. For more information, see
   :ref:`the comparison <comparison-to-the-os-and-os-path-modules>`.
```

**Comment by SubaruArai** [comment]
Oh, I forgot - thank you very much for tackling this PR!
As for the Qs:

> The pathlib docs are pretty beginner-friendly, so I went back to basics with the OOP vs procedural stuff. Is it too much?

I think it's fine.
From a practical point of view(see below), I'd rather have this after the functional differences.
But then again, if I was a beginner I would want to read this first as an introduction so I guess it's down to personal preference.

I.e. when I'm reading it, I'm looking for why should I choose `os.path` where everyone is saying to use `pathlib`.

> Is the trailing slash bit clear enough?

I think so.

> Should the conclusion "pathlib is not a drop-in replacement" be softened?

I think it's fine. For me, the main target audience for this line are package maintainers that are thinking the need to refactor. I feel they need to keep this in mind.

**Comment by tovrstra** [comment]
Well written. Thanks for taking care of the clarification.

I would have appreciated a reference to [path](https://github.com/jaraco/path), e.g. in a footnote, but if it is considered out of scope, fine by me.

As an end-user, I found it a pity to discover this "third" option rather late, and I prefer it over pathlib, because it gives the user control over when normalization happens, while still offering the convenience of an object-oriented interface. Given the elaboration here and discussions in [topic: leading dot](https://discuss.python.org/t/suggestion-for-pathlib-differentiate-explicit-and-implicit-local-paths-pathlib-strictpath/31235) [topic: trailing slash](https://discuss.python.org/t/pathlib-preserve-trailing-slash/33389), it seems justified to add a helpful suggestion for those who are unhappy with the built-in options.

**Comment by SubaruArai** [comment]
@tovrstra I thought about adding `path` too (thanks for pointing out in the parent issue!).
The reason I was reluctant was:

1. I don't have first-hand experience using it
2. a quick search showed very little mentioning about 3rd party modules. numpy and scipy [got mentioned](https://docs.python.org/3/library/statistics.html) though.
3. Ideally I wanted another 3rd party library, to give users some choice.

But the majority of reason was 1. I don't have experience on it enough (incl. the project's management record) to recommend to someone, let alone showing it as the sole recommendation for better alternatives on the official python docs.

**Comment by barneygale** [comment]
I thought about recommeding `path`, but I'm not sure about it inheriting from `str`. Obviously it long pre-dates `os.PathLike` so I understand why it does.

**Comment by tovrstra** [comment]
I'm fine either way (referring `path` or not). If more 3rd-party projects are of interest: [PyFilesystem](https://www.pyfilesystem.org/) might be worth mentioning. It is slightly outdated and the scope also differs more from `pathlib`, so one could question its relevance here.

@barneygale Subclassing from `str` is reasonable, I think. One of the advantages is that existing libraries that accept paths as strings will work without modifications when given a `path.Path` instance. I can see there are also reasons not to subclass `str`, so it's a matter of trade-offs, which may be subject to the use case.

**Comment by barneygale** [comment]
Thanks for the feedback + reviews, both. Very much appreciated.

I'll exclude the reference to the `path` package. The new section describes, basically, the various deficiencies of `pathlib` and `os.path`. It wouldn't be fair to recommend `path` uncritically in the same section, as I'd imagine it has its own deficiencies, and I'm not familiar enough with `path` to comprehend and describe those deficiencies. I hope that makes sense.

**Pull request merged**